connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectStreamingQueryCache.scala

```diff
@@ -0,0 +1,200 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connect.service
+
+import java.util.concurrent.Executors
+import java.util.concurrent.ScheduledExecutorService
+import java.util.concurrent.TimeUnit
+import javax.annotation.concurrent.GuardedBy
+
+import scala.collection.mutable
+import scala.concurrent.duration.Duration
+import scala.concurrent.duration.DurationInt
+import scala.util.control.NonFatal
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.streaming.StreamingQuery
+import org.apache.spark.util.Clock
+import org.apache.spark.util.SystemClock
+
+/**
+ * Caches Spark-Connect streaming query references and the sessions. When a query is stopped (i.e.
+ * no longer active), it is cached for 1 hour so that it is accessible from the client side. It
+ * runs a background thread to run a periodic task that does the following:
+ *   - Check the status of the queries, and drops those that expired (1 hour after being stopped).
+ *   - Keep the associated session active by invoking supplied function `sessionKeepAliveFn`.
+ *
+ * This class helps with supporting following semantics for streaming query sessions:
+ *   - Keep the session and session mapping at connect server alive as long as a streaming query
+ *     is active. Even if the client side has disconnected.
+ *     - This matches how streaming queries behave in Spark. The queries continue to run if
+ *       notebook or job session is lost.
+ *   - Once a query is stopped, the reference and mappings are maintained for 1 hour and will be
+ *     accessible from the client. This allows time for client to fetch status.
+ *     - During this time if the query is restarted (i.e. has a new run id), the reference to
+ *       previous run is dropped. As a result logical query has only the most recent query
+ *       reference cached. This policy can be revisited to cache multiple runs for a query.
+ *
+ * Note that these semantics are evolving and might change before being finalized in Connect.
+ */
+private[connect] class SparkConnectStreamingQueryCache(
+    val sessionKeepAliveFn: (String, String) => Unit, // (userId, sessionId) => Unit.
+    val clock: Clock = new SystemClock(),
+    private val stoppedQueryCachePeriod: Duration = 1.hour, // Configurable for testing.
+    private val sessionPollingPeriod: Duration = 1.minute // Configurable for testing.
+) extends Logging {
+
+  import SparkConnectStreamingQueryCache._
+
+  def registerNewStreamingQuery(sessionHolder: SessionHolder, query: StreamingQuery): Unit = {
```

Comment:
The possible problematic case is that multiple remote sessions are trying to run the streaming query with same checkpoint. Pretty sure it's not something Spark supports and one of the query will eventually survive (others will fail), but at least driver is able to run multiple streaming queries with the same checkpoint concurrently. Please make sure this won't break the key of the cache. Using runId would avoid the problem, but I guess there would be a reason to use query ID instead.

Reply:
Good point. We only keep track of most recent run intentionally (also mentioned in the semantics). We can keep track of all the runs, but that might increase number of queries tracked (especially if there is a query in crashloop or trigger.once() loop). 
There is nothing technically blocking us to track at runId level.

So in the above scenario, which run ends up seen last will survive. The rest of them will not be tracked.

Reply:
We discussed in offline but summarizing here for other reviewers / future reference:

There is inconsistency between selecting winner. In cache mechanism, later one always wins the race and overwrites the old one. In concurrent query execution, we never know which one wins.

For example, say there are three users userA userB and userC. They somehow (wrongly) refer the same checkpoint and run the query at the same time. (Three query with different run IDs but same unique query ID would exist at the same time.) While there will be eventually only one winner, individual should be able to track their own query status, despite that they are determined as loser.

Reply:
I will update the PR to cache all the instances of a query (ie. key is `query_id, run_id`).  This will handle various scenarios better.

One reason for not starting with this in first pass was to limit queries cached in case of crash loop of query. We can handle that (not in this PR for now).

Reply:
Update the PR with two changes:
  - Include `runId` in the cache. As a result, each run of a query will be maintained independently (unit test is updated).
  - After a query is stopped, it is cached for 1 hour. But if the client keeps accessing the query, will keep extending this timeout. I.e. 1 hour is an inactivity timeout. 

@pengzhon-db PTAL. @HeartSaVioR will be out on Monday too due to public holiday in SK. We can get this merged and address any remaining comments as a follow up.

Reply:
LGTM